✨ feat: add Github Models provider

CloudPassenger opened the pull request:
#### 💻 变更类型 | Change Type

<!-- For change type, change [ ] to [x]. -->

- [x] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 👷 build
- [ ] ⚡️ perf
- [ ] 📝 docs
- [ ] 🔨 chore

#### 🔀 变更说明 | Description of Change

Support [Github Models](https://docs.github.com/en/github-models/prototyping-with-ai-models) as LLM model provider

#### 📝 补充信息 | Additional Information

- [x] Add `GithubModel` to `ProviderCombine` in `@lobehub/icons`

arvinxx commented on src/server/routers/lambda/ragEval.ts:
```diff
@@ -212,7 +212,7 @@ export const ragEvalRouter = router({
 
                 throw new TRPCError({
-                  code: 'BAD_GATEWAY',
+                  code: 'INTERNAL_SERVER_ERROR',
```
这个应该是 BAD_GATEWAY ，而不是 INTERNAL_SERVER_ERROR

CloudPassenger replied:
这个的话如果保持不修改，会出现这样一个奇怪的错误，导致我不能commit

```
Type '"BAD_GATEWAY"' is not assignable to type '"BAD_REQUEST" | "PARSE_ERROR" | "INTERNAL_SERVER_ERROR" | "NOT_IMPLEMENTED" | "UNAUTHORIZED" | "FORBIDDEN" | "NOT_FOUND" | "METHOD_NOT_SUPPORTED" | "TIMEOUT" | "CONFLICT" | ... 5 more ... | "CLIENT_CLOSED_REQUEST"'.
```

不知道是不是我这边的依赖问题

arvinxx replied:
你升级下依赖试试？

CloudPassenger committed:
✨ feat: add Github Models as provider

canisminor1990 commented:
- [x] Add `GithubModel` to `ProviderCombine` in `@lobehub/icons`

CloudPassenger commented:
> @CloudPassenger rebase 下main， ProviderAvatar 这个收到组件里去了

已经 rebase，麻烦查阅了，谢谢

arvinxx commented:
<img width="729" alt="image" src="https://github.com/user-attachments/assets/e2dee2f7-e01c-4651-a5b0-5071e2bbd6c6">

确定一下是用 Token 还是 API Key

CloudPassenger commented:
> <img alt="image" width="729" src="https://private-user-images.githubusercontent.com/28616219/366478679-e2dee2f7-e01c-4651-a5b0-5071e2bbd6c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNzE3NzAsIm5iZiI6MTcyNjA3MTQ3MCwicGF0aCI6Ii8yODYxNjIxOS8zNjY0Nzg2NzktZTJkZWUyZjctZTAxYy00NjUxLWE1YjAtNTA3MWUyYmJkNmM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDE2MTc1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlYjU2YjM0OWE3ZTVkZDk0MDE4ZjAyNGU0N2NjNjJkNjcwYjkzZTEzNmEwMTIzZDdjY2QwY2JhNmM4OWIxNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ohu3WE-0CIk3KGnaFgp--Cl7fPWFAf5ebE7oQ5q3PPo">
> 确定一下是用 Token 还是 API Key

https://github.com/lobehub/lobe-chat/blob/38f6c345d9eb1eb43204535aa44e674843021d21/src/app/(main)/settings/llm/components/ProviderConfig/index.tsx#L147-L161

是 Github Token，但我查看了一下并没有给不同 Provider 配置显示的选项，如果我在这里添加根据 id 显示不同文字的逻辑，并且更新 i18n 的话，会是一个好的实践吗？

arvinxx commented:
> 如果我在这里添加根据 id 显示不同文字的逻辑，并且更新 i18n 的话，会是一个好的实践吗？

这个方案不太好，建议是参考 Azure 或者 Bedrock 的方案，外部覆写下 `apiKeyItems` ：

https://github.com/lobehub/lobe-chat/blob/38f6c345d9eb1eb43204535aa44e674843021d21/src/app/(main)/settings/llm/ProviderList/Bedrock/index.tsx#L19-L30

针对 Github 这个再单独加 i18n 比较好

LovelyGuYiMeng commented:
![QQ20240912-110258](https://github.com/user-attachments/assets/5d5767cc-981c-4a3a-baf3-cb34bffc966a)
![QQ20240912-110327](https://github.com/user-attachments/assets/a345678e-23e7-4690-a064-5324815f7bd5)
请注意排序务必相同，不要一个在Deepseek后面一个在Openrouter后面

CloudPassenger committed:
:bug fix: sort order of ProviderList

CloudPassenger committed:
💄 style: add more info for Github Models provider

CloudPassenger committed:
✨ feat: Add GitHub provider configuration

Adds support for configuring the GitHub provider in the LLM settings. It includes:

- A new `useGithubProvider` hook to define the GitHub provider configuration
- Integration of the GitHub provider in the provider list
- Localization strings for the GitHub personal access token input

The changes allow users to enter their GitHub personal access token to use GitHub-related LLM features.

CloudPassenger committed:
🐛 fix: use Github Personal access tokens (classic) instead

CloudPassenger committed:
💄 style: Add Invalid Github Token error type and message

Adds support for handling invalid Github Personal Access Tokens:

- Adds a new `InvalidGithubToken` error type in the `AgentRuntimeErrorType` enum
- Adds an error message for `InvalidGithubToken` in the localization files
- Configures the Github AI provider to use the new error type

This allows properly handling and displaying errors when an invalid Github token is provided.

CloudPassenger committed:
✅ test: fix test for new error type

CloudPassenger commented:
I have made some changes for this feature. Please review.
If you have any other suggestions, please don't hesitate to let me know. I will get back to you as soon as possible. Thanks!

LovelyGuYiMeng commented on src/app/(main)/settings/llm/ProviderList/providers.tsx:
```diff
@@ -44,9 +46,10 @@ export const useProviderList = (): ProviderItem[] => {
       DeepSeekProviderCard,
       OpenRouterProviderCard,
+      GithubProvider,
+      GoogleProviderCard,
```
把GoogleProviderCard放到BedrockProviderCard后面，不要乱改顺序

CloudPassenger replied:
不好意思，已经改正

LovelyGuYiMeng reviewed (commented):
把GoogleProviderCard放到BedrockProviderCard后面，不要乱改顺序

CloudPassenger committed:
🐛 fix: sort order of ProviderCards

arvinxx commented:
1. Rebase 下
2. 改成缩写吧？ GitHub PAT ，不然看着太长了，被裁剪了。 然后 desc 那边最好加个超链接直接跳创建 Github PAT 的地方？

<img width="693" alt="Google Chrome 2024-09-13 23 31 59" src="https://github.com/user-attachments/assets/7ee23247-c5ef-41f3-9f63-4edbc2826885">

CloudPassenger committed:
💄 style: use abbreviation for Github PAT

CloudPassenger committed:
💄 style: add the hyperlink for creating Github PAT

CloudPassenger commented:
> @CloudPassenger
> 
> 1. Rebase 下
> 2. 改成缩写吧？ GitHub PAT ，不然看着太长了，被裁剪了。 然后 desc 那边最好加个超链接直接跳创建 Github PAT 的地方？
> 
> <img alt="Google Chrome 2024-09-13 23 31 59" width="693" src="https://private-user-images.githubusercontent.com/28616219/367356807-7ee23247-c5ef-41f3-9f63-4edbc2826885.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0ODgxMzcsIm5iZiI6MTcyNjQ4NzgzNywicGF0aCI6Ii8yODYxNjIxOS8zNjczNTY4MDctN2VlMjMyNDctYzVlZi00MWYzLTlmNjMtNGVkYmMyODI2ODg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDExNTcxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3MDMyODM2MzZiNmE2MTJmZGJlYzUyN2IxNDU5YjBmYzk2ZmFkYzBlODI3YmYzMmM1ZjlhODg3NDA4YWMxYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sK4V3vSV7MzQwvuk0maRmHa0lsfJvuDrIjoxqSjNXTY">

感谢提醒，已经修复并添加，中秋快乐🥮

arvinxx committed:
🌐 chore: improve desc

arvinxx committed:
Merge branch 'main' into provider/github-models

arvinxx committed:
✅ test: update test
